### PR TITLE
Consolidate QMCHamiltonian code paths.

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Rename `updateNonKinetic` to `updateComponent` and adjust `updateKinetic`. Maximize code reuse.

## What type(s) of changes does this code introduce?
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

**Review by PDoakORNL — COMMENTED** [review]
Looking at this now.

File: src/QMCHamiltonians/QMCHamiltonian.cpp [hunk 1]
```diff
@@ -572,34 +560,22 @@ QMCHamiltonian::FullPrecRealType QMCHamiltonian::evaluateDeterministic(ParticleS
...
 #if !defined(REMOVE_TRACEMANAGER)
     H[i]->collectScalarTraces();
 #endif
-    H[i]->setParticlePropertyList(P.PropertyList, myIndex);
   }
-  KineticEnergy                      = H[0]->getValue();
-  P.PropertyList[WP::LOCALENERGY]    = LocalEnergy;
-  P.PropertyList[WP::LOCALPOTENTIAL] = LocalEnergy - KineticEnergy;
-  // auxHevaluate(P);
+  updateKinetic(*this, P);
```
Comment (PDoakORNL) [inline src/QMCHamiltonians/QMCHamiltonian.cpp:569]: I wish there was a unit test here and elsewhere to check for instance that the LocalEnergy value is indeed the same, but this refactor looks ok by instpection.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
QMCHamiltonian could really use a unit test to either confirm the local energies the methods return are correct/consistent with those coming via the property lists. This would then be simple to approve.  But as it stands this is a refactor that is missing the unit test it should depend on.

File: src/QMCHamiltonians/QMCHamiltonian.cpp (hunk 1 above)
Comment (ye-luo) [reply src/QMCHamiltonians/QMCHamiltonian.cpp:569]: Did you mean testing
`P.PropertyList[WP::LOCALENERGY]   == ham.getLocalEnergy()` ?

Comment (ye-luo) [reply src/QMCHamiltonians/QMCHamiltonian.cpp:569]: I added checks in one of the unit test.

File: src/QMCHamiltonians/tests/test_QMCHamiltonian.cpp [hunk 2]
```diff
@@ -203,6 +204,8 @@ TEST_CASE("integrateListeners", "[hamiltonian]")
 
   QMCHamiltonian::mw_evaluate(ham_list, twf_list, p_list);
```
Comment (PDoakORNL) [inline src/QMCHamiltonians/tests/test_QMCHamiltonian.cpp:205]: this call also returns a std::vector of local energy which should match. 

And the sum of the per particle component energies and the sum of local energies per particle sum should match this returned value as well. (This is a test I should have added so you can just drop an issue for me.)

Comment (ye-luo) [reply src/QMCHamiltonians/tests/test_QMCHamiltonian.cpp]: > this call also returns a std::vector of local energy which should match.

I added checks for this.

> And the sum of the per particle component energies and the sum of local energies per particle sum should match this returned value as well. (This is a test I should have added so you can just drop an issue for me.)

Added https://github.com/QMCPACK/qmcpack/issues/4769

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

**Review by PDoakORNL — APPROVED** [review]
LGTM

**Pull request merged**